### PR TITLE
New chat events for better Adventure support

### LINKED ISSUE #5216: "Adventure" Chat issues.
### What behaviour is expected: 
I expected no changes to chat formatting.

### What behaviour is observed:
Opped players now have 2 :'s after their username. Essentials Social Spy now is written in the color of the players nickname.
image

### Steps/models to reproduce:
All I did was update paper about 7 versions. 

### Plugin list:
image

### Paper version:
image

### LINKED ISSUE #5261: Stray §r added to legacy text where unnecessary and unexpected
### What behaviour is expected:

The following printed to the console: `&a&lP&eaper`.

### What behaviour is observed:

The following printed to the console: `&a&lP&r&eaper`.

### Steps/models to reproduce:

```kotlin
val state = block.state as Sign
val line = state.getLine(0).replace('§', '&')
Bukkit.broadcastMessage(line)
```

Have a sign whose first line is set to the legacy text: `&a&lP&eaper` and run the above code on it.

### Plugin list:

```
Plugins (16): BlockLocker, ChatItemDisplay, Essentials, EssentialsSpawn, Gaea, grakkit*, LuckPerms, PlaceholderAPI, ProtocolLib, ProxiDebugPlugin, ServerUtils, spark, Stress, Terra, Vault, WorldEdit
```

### Paper version:

```
[19:16:25 INFO]: This server is running Paper version git-Paper-"fc072283c" (MC: 1.16.5) (Implementing API version 1.16.5-R0.1-SNAPSHOT)
[19:16:25 INFO]: Checking version, please wait...
[19:16:25 INFO]: Previous version: git-Paper-498 (MC: 1.16.5)
[19:16:25 INFO]: Error obtaining version information
```

### LINKED ISSUE #5287: Chat format rendering is inconsistent between Paper and Spigot (Adventure)
### What behaviour is expected:
Consistent formatting between chat formats on Spigot and Paper.

On spigot: after a display name or chat message is replaced the the formatting after the message remains the same.

For example, the format, `&e<%1$s> &d%2$s |`, yields the following on Spigot

display name of `&#63320c~Josh&r` and message `t`:
image

display name of `&#63320c~Josh&r` and message `&ey`:
image

### What behaviour is observed:

On Paper, with the same format as above:

display name of `&#63320c~Josh&r` and message `t`:
image

display name of `&#63320c~Josh&r` and message `&ey`:
image

It appears that adventure is replacing the display name and message in the chat format late and therefore isolating the format of the display name of message. While I think this is better and how it should have been, this breaks compatibility with Spigot unfortunately and causes problems with chat plugins :/

### Steps/models to reproduce:
Follow the scenarios list above. I used Essentials and EssentialsChat to do this, if you want to go that route, chat formats are @ `chat.format` in the config and you can change your display name with `/nick`

### Plugin list:
```
Essentials, EssentialsChat, EssentialsProtect, EssentialsSpawn, LuckPerms, spark, Vault
```

### Paper version:
```
[09:34:37 INFO]: This server is running Paper version git-Paper-505 (MC: 1.16.5) (Implementing API version 1.16.5-R0.1-SNAPSHOT)
[09:34:37 INFO]: Checking version, please wait...
[09:34:37 INFO]: Previous version: git-Paper-498 (MC: 1.16.5)
[09:34:37 INFO]: You are running the latest version
```

### PR DESCRIPTION
**Pull request by kashike**

### CODE REVIEW

File: Spigot-Server-Patches/0010-Adventure.patch [hunk 1]
```diff
@@ -1199,7 +1394,22 @@ index d34e91887cd73009bf852fb849e495a8affed7a9..5fa4afb75d46a32d91402c5ca850d17d
...
+         } else if (this.player.getChatFlags() == EnumChatVisibility.SYSTEM) {
+             // Do nothing, this is coming from a plugin
+-        } else {
+-            Player player = this.getPlayer();
++        } else if (true) {
++            final ChatProcessor cp = new ChatProcessor(this.minecraftServer, this.player, s, async);
++            cp.process();
++        // Paper end
++        } else if (false) { // Paper
++            CraftPlayer player = this.getPlayer(); // Paper
```
Comment (MiniDigger) [inline Spigot-Server-Patches/0010-Adventure.patch]: why did you change this line if its unreachable?

File: Spigot-API-Patches/0005-Adventure.patch [hunk 2]
```diff
@@ -828,6 +1048,19 @@ index fcc34b640265f4dccb46b9f09466ab8e1d96043e..74599b4ee0518481c0e3a5f6ab2f5302
...
+diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
+index f020cb04eba27a2e70fc7cf799ebbfb434b9d974..8ad25e86498c56d7ca8d118fcedd194f8daf1824 100644
+--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
++++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
+@@ -66,7 +66,7 @@ public class SimpleCommandMap implements CommandMap {
+      */
+     @Override
+     public boolean register(@NotNull String label, @NotNull String fallbackPrefix, @NotNull Command command) {
+-        command.timings = co.aikar.timings.TimingsManager.getCommandTiming(fallbackPrefix, command); // Paper
++//        command.timings = co.aikar.timings.TimingsManager.getCommandTiming(fallbackPrefix, command); // Paper
```
Comment (MiniDigger) [inline Spigot-API-Patches/0005-Adventure.patch]: forgot to revert this or why was this disabled?

File: Spigot-Server-Patches/0010-Adventure.patch [hunk 3]
```diff
@@ -104,6 +104,200 @@ index 0000000000000000000000000000000000000000..a2acd31dce4461338a8baa96e03df36a
...
++        final Consumer<AsyncPlayerChatEvent> continueAfterAsync,
++        final Consumer<PlayerChatEvent> continueAfterAsyncAndSync,
++        final Runnable modernOnly
++    ) {
++        final boolean listenersOnAsyncEvent = anyListeners(AsyncPlayerChatEvent.getHandlerList());
++        final boolean listenersOnSyncEvent = anyListeners(PlayerChatEvent.getHandlerList());
++        if (listenersOnAsyncEvent || listenersOnSyncEvent) {
++            final CraftPlayer player = this.player.getBukkitEntity();
++            final AsyncPlayerChatEvent ae = new AsyncPlayerChatEvent(this.async, player, this.message, new LazyPlayerSet(this.server));
++            post(ae);
```
Comment (MiniDigger) [inline Spigot-Server-Patches/0010-Adventure.patch:279]: can just do ae.callEvent(), no need for the post method

**Review by MiniDigger — COMMENTED** [review]
logic looks fine to my mushed brain, needs testing with all possible combinations I guess

**Comment by mrfloris** [comment]
This update is causing CMI Chat to show up double in-game. Build 506 is okay, build 507 shows double.
